Add new cmdlet, update commands.Sql and tests for 'List database operation' and 'List/Cancel elastic pool operation'

payiAzure opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description

The Changes in the PR is made from local .NET SDK binaries. Will update this PR after the corresponding .NET SDK is published.

The changes included in the PR:
  1. Add new cmdlet: Stop-AureRmSqlElasticPoolActivity -ResourceGroupName <rgName> -ServerName <serverName> -ElasticPoolName <poolName> -OperationId <operationId>

  2. Update the AzureSqlDatabaseActivityModel and AzureSqlElasticPoolActivityModel. When user try to get the database/elasticPool operations, the returned information of the operation will also include the new fields: 
      estimated completion time of the operation; 
      detail operation description;
      mark to indicate if the operation can be cancelled.

3. Add test for the new added cmdlet.

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#making-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

jaredmoo commented on src/ResourceManager/Sql/Commands.Sql/Elastic Pools/Services/AzureSqlElasticPoolAdapter.cs:
```diff
@@ -164,12 +165,61 @@ internal ICollection<AzureSqlDatabaseModel> ListElasticPoolDatabases(string reso
 
-            return resp.Select((activity) =>
+            var resp = from activity in activityResp
+                       join operation in operationResp
```
Nice!

praries880 commented on src/ResourceManager/Sql/Commands.Sql/Elastic Pools/Services/AzureSqlElasticPoolAdapter.cs:
```diff
@@ -189,6 +239,27 @@ internal IList<AzureSqlDatabaseActivityModel> ListElasticPoolDatabaseActivity(st
         }
 
+        /// <summary>
+        /// Cancel teh elastic pool activity
```
Typo : "teh"

praries880 commented on src/ResourceManager/Sql/Commands.Sql/Elastic Pools/Cmdlet/StopAzureSqlElasticPoolActivity.cs:
```diff
@@ -0,0 +1,54 @@
+        /// No user input to apply to model
+        /// </summary>
+        /// <param name="model">Model retrieved from service</param>
+        /// <returns>The model that wa passed in</returns>
```
Typo : "wa"

praries880 commented on src/ResourceManager/Sql/Commands.Sql/help/Stop-AzureRmSqlElasticPoolActivity.md:
```diff
@@ -0,0 +1 @@

## NOTES

## RELATED LINKS
```
Synopsis is empty : "{{Fill in the Synopsis}}"

cormacpayne commented on src/ResourceManager/Sql/Commands.Sql/Elastic Pools/Cmdlet/StopAzureSqlElasticPoolActivity.cs:
```diff
@@ -0,0 +1,54 @@
+
+namespace Microsoft.Azure.Commands.Sql.ElasticPool.Cmdlet
+{
+    [Cmdlet(VerbsLifecycle.Stop, "AzureRmSqlElasticPoolActivity", SupportsShouldProcess = true)]
```
@payiAzure please add an `OutputType` to this cmdlet -- if this cmdlet does not return any output, then you should implement the [`PassThru`](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/azure-powershell-design-guidelines.md#returning-no-output) parameter

payiAzure replied:
@cormacpayne I'm not clear about what is the PassThru parameter and how to implement it? can you please point me to any example?
Since I just follow one existing command to implement this similar one. "StopAzureSqlDatabaseActivity.cs"

payiAzure replied:
@cormacpayne .Add OutputType and implement the PassThru parameter.

cormacpayne commented on src/ResourceManager/Sql/Commands.Sql.Test/ScenarioTests/ElasticPoolCrudTests.ps1:
```diff
@@ -297,4 +297,57 @@ function Test-RemoveElasticPool
+function Test-ListAndCancelElasticPoolOperation
+{
+	# Setup
+	$location = "southeastasia";
```
@payiAzure please avoid hard-coding locations in tests -- this potentially makes it difficult for other users to re-record this test who might not have access to the resources in the provided location. We recommend using the [`Get-Location`](https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/Common/Commands.ScenarioTests.ResourceManager.Common/Common.ps1#L583-L604) function in our test common code to get a location that works for the user re-recording the test.

cormacpayne commented on src/ResourceManager/Sql/Commands.Sql.Test/ScenarioTests/ElasticPoolCrudTests.ps1:
```diff
@@ -297,4 +297,57 @@ function Test-RemoveElasticPool
+
+		try
+		{
+			$activityCancel = Stop-AzureRmSqlElasticPoolActivity -ResourceGroupName $ep.ResourceGroupName -ServerName $ep.ServerName -ElasticPoolName $ep.ElasticPoolName -OperationId $epactivityId
```
@payiAzure just to be sure, should this command _always_ error out? If not, then we should add some asserts over the returned `$activityCancel` object

payiAzure replied:
yes, in the test, it will always error out since the previous update operation makes the pool in an state that is not available in cancel.

payiAzure commented on src/ResourceManager/Sql/Commands.Sql/help/Stop-AzureRmSqlElasticPoolActivity.md:
```diff
@@ -0,0 +1 @@

## NOTES

## RELATED LINKS
```
@cormacpayne updated this md file, should have correct format now.

payiAzure commented on src/ResourceManager/Sql/Commands.Sql/Elastic Pools/Cmdlet/StopAzureSqlElasticPoolActivity.cs:
```diff
@@ -0,0 +1,54 @@
+namespace Microsoft.Azure.Commands.Sql.ElasticPool.Cmdlet
+{
+    [Cmdlet(VerbsLifecycle.Stop, "AzureRmSqlElasticPoolActivity", SupportsShouldProcess = true)]
+
```
@cormacpayne removed in the new commit.

jaredmoo commented on src/ResourceManager/Sql/AzureRM.Sql.psd1:
```diff
@@ -189,7 +189,8 @@ CmdletsToExport = 'Get-AzureRmSqlDatabaseTransparentDataEncryption',
                'Remove-AzureRmSqlServerDnsAlias', 'New-AzureRmSqlServerDnsAlias', 
-               'Set-AzureRmSqlServerDnsAlias'
+               'Set-AzureRmSqlServerDnsAlias',
+			   'Stop-AzureRmSqlElasticPoolActivity'
```
nit: fix indentation here

jaredmoo commented on src/ResourceManager/Sql/Commands.Sql/Elastic Pools/Services/AzureSqlElasticPoolAdapter.cs:
```diff
@@ -189,6 +239,27 @@ internal IList<AzureSqlDatabaseActivityModel> ListElasticPoolDatabaseActivity(st
+            Communicator.CancelOperation(resourceGroupName, serverName, elasticPoolName, operationId.Value);
+
+            // After Cancel event is fired, state will be in 'CancelInProgress' for a while but should expect to finish in a minute
+            return GetElasticPoolActivity(resourceGroupName, serverName, elasticPoolName);
```
nice!

jaredmoo commented on src/ResourceManager/Sql/Commands.Sql/help/Stop-AzureRmSqlElasticPoolActivity.md:
```diff
@@ -0,0 +1,171 @@
+# Stop-AzureRmSqlElasticPoolActivity
+
+## SYNOPSIS
+Cancels the asynchronous updates operation on the elastic pool.
```
`Cancels an asynchronous update operation on an elastic pool.`

jaredmoo commented on src/ResourceManager/Sql/Commands.Sql/help/Stop-AzureRmSqlElasticPoolActivity.md:
````diff
@@ -0,0 +1,171 @@
+```
+
+## DESCRIPTION
+The **Stop-AzureRmSqlElasticPoolActivity** cmdlet cancels the asynchronous updates operation on the elastic pool.
````
`... cancels an asynchronous update operation on an elastic pool.`

jaredmoo commented on src/ResourceManager/Sql/Commands.Sql/help/Stop-AzureRmSqlElasticPoolActivity.md:
````diff
@@ -0,0 +1,171 @@
+PercentComplete : 
+```
+
+This command cancels the asynchronous updates operation on the elastic pool.
````
same comment

payiAzure replied:
@jaredmoo , comments addressed in the new commit.

jaredmoo commented on src/ResourceManager/Sql/Commands.Sql/help/Stop-AzureRmSqlElasticPoolActivity.md:
````diff
@@ -0,0 +1,171 @@
+
+### Example 1: Cancel the asynchronous updates operation on the elastic pool
+```
+PS C:\> Stop-AzureRmSqlElasticPoolActivity -ResourceGroupName "ResourceGroup01" -ServerName "Server01" -ElasticPoolName "ElasticPool01" -OperationId af97005d-9243-4f8a-844e-402d1cc855f5
````
quotes not needed

cormacpayne commented on src/ResourceManager/Sql/Commands.Sql/help/Stop-AzureRmSqlElasticPoolActivity.md:
````diff
@@ -0,0 +1,171 @@
+### Example 1: Cancel the asynchronous update operation on an elastic pool
+```
+PS C:\> Stop-AzureRmSqlElasticPoolActivity -ResourceGroupName "ResourceGroup01" -ServerName "Server01" -ElasticPoolName "ElasticPool01" -OperationId af97005d-9243-4f8a-844e-402d1cc855f5
+```
````
@payiAzure please remove the three backticks on this line

payiAzure replied:
@cormacpayne removed in the new commit.

cormacpayne commented on src/ResourceManager/Sql/Commands.Sql/help/AzureRM.Sql.md:
```diff
@@ -407,6 +407,9 @@ Cancels the asynchronous updates operation on the database.
 Stops the workflow that runs a recommended index operation.
 
+### [Stop-AzureRmSqlElasticPoolActivity](Stop-AzureRmSqlElasticPoolActivity.md)
+{{Fill in the Synopsis}}
```
@payiAzure please fill in this synopsis

payiAzure replied:
@cormacpayne add synopsis in the AzureRmSql.md for the new command.

cormacpayne commented on src/ResourceManager/Sql/ChangeLog.md:
```diff
@@ -24,6 +24,8 @@
 ## Version 4.3.0
 * Fixed issue with importing aliases
 * Get-AzureRmSqlServer, New-AzureRmSqlServer, and Remove-AzureRmSqlServer response now includes FullyQualifiedDomainName property.
+* Add new cmdlet 'Stop-AzureRmSqlElasticPoolActivity' to support canceling the asynchronous operations on elastic pool
```
@payiAzure would you mind pulling the latest from the `preview` branch? The content you are adding should be under the `## Current Release` header

payiAzure replied:
@cormacpayne pull the latest from preview branch and put my changes under the ##Current Release header

cormacpayne commented on src/ResourceManager/Sql/Commands.Sql.Test/ScenarioTests/ElasticPoolCrudTests.ps1:
```diff
@@ -297,4 +297,91 @@ function Test-RemoveElasticPool
+		try
+		{
+			# cancel a pool update operation using piping
+			$activityCancel = $epactivity | Stop-AzureRmSqlElasticPoolActivity
```
@payiAzure since we always expect this to error out, we should change this to the following:

```powershell
Assert-Throws { $ep2update | Get-AzureRmSqlElasticPoolActivity -OperationId $epactivityId }
```

cormacpayne commented on src/ResourceManager/Sql/Commands.Sql/Commands.Sql.csproj:
```diff
@@ -224,6 +225,7 @@
       <Link>AzureRM.Sql.psd1</Link>
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
+    <None Include="app.config" />
```
@payiAzure please remove this reference

payiAzure replied:
@cormacpayne removed.

cormacpayne commented on src/ResourceManager/Sql/Commands.Sql/Commands.Sql.csproj:
```diff
@@ -447,18 +449,41 @@
+    <Reference Include="Microsoft.Azure.Management.Sql, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Sql.1.13.0-preview\lib\net452\Microsoft.Azure.Management.Sql.dll</HintPath>
+    </Reference>
+    <!--<Reference Include="Microsoft.Azure.Management.Sql, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
@payiAzure please remove this commented out section

payiAzure replied:
@cormacpayne removed.

cormacpayne commented on src/ResourceManager/Sql/Commands.Sql/Commands.Sql.csproj:
```diff
@@ -447,18 +449,41 @@
     <Reference Include="Microsoft.Azure.Management.Storage">
       <HintPath>..\..\..\packages\Microsoft.Azure.Management.Storage.2.4.0-preview\lib\net40\Microsoft.Azure.Management.Storage.dll</HintPath>
     </Reference>
+    <Reference Include="Microsoft.Data.Edm, Version=5.8.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
@payiAzure there are a bunch of references being added to this file that aren't needed -- the only change that needs to be made in this section of the file is the version bump of `Microsoft.Azure.Management.Sql` to version `1.13.0-preview` and the removal of the commented out reference. All other changes can be removed.

payiAzure replied:
@cormacpayne removed other un-necessary references.
Microsoft.Data.Edm
Microsoft.Azure.KeyVault.Core
Microsoft.Data.Services.Client
Microsoft.Data.OData
Microsoft.Rest.ClientRuntime
Microsoft.Rest.ClientRuntime.Azure
Microsoft.WindowsAzure.Storage, Version=8.1.1.0
Newtonsoft.Json
System.Configuration.Install

cormacpayne commented on src/ResourceManager/Sql/Commands.Sql/help/Stop-AzureRmSqlElasticPoolActivity.md:
```diff
@@ -0,0 +1,170 @@
+﻿---
+external help file: Microsoft.Azure.Commands.Sql.dll-Help.xml
+Module Name: AzureRM.Sql
+online version:
```
@payiAzure the build will fail because there is no online help URL found here. It should be `https://docs.microsoft.com/en-us/powershell/module/azurerm.sql/stop-azurermsqlelasticpoolactivity`

payiAzure replied:
@cormacpayne adding the help URL

cormacpayne commented on src/ResourceManager/Sql/Commands.Sql/help/Stop-AzureRmSqlElasticPoolActivity.md:
```diff
@@ -0,0 +1,170 @@
+
+## INPUTS
+
+### System.String
```
@payiAzure change this `INPUTS` section to the following:

```
## INPUTS

### None

You can pipe input to this cmdlet by property name, but not by value.
```

payiAzure replied:
@cormacpayne added in the next commit

payiAzure commented:
@cormacpayne Can you please review this again? I have addressed the comments. I add the help url in the help file, but not sure why the build check still fail here. I can build from local.

cormacpayne commented:
@payiAzure would you mind taking a look at the merge conflicts? Also, the build is failing because the `packages.config` files in the Sql projects have not been updated with the new version of `Microsoft.Azure.Management.Sql`. Please make sure to update these files when pulling in the latest changes from the `preview` branch.

payiAzure commented:
@cormacpayne Fix the merge conflict issue and the build failure issues. Could you please take a review? Thank you!

cormacpayne commented:
On-demand: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/675/

maddieclayton commented:
rerun on demand here: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/683/

markcowl commented:
new on-demand run here: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/684/
